valueset.go

```diff
@@ -0,0 +1,261 @@
+	processed := false
+	ret := make(map[string]Nodeset)
+	getColumn := func(row []string, header string) string {
+		for cIdx, c := range input.ColumnNames() {
+			if c == header && cIdx < len(row) {
+				return row[cIdx]
+			}
+		}
+		return ""
+	}
+	for {
+		if err := input.Reset(); err != nil {
+			return nil, err
+		}
+		var currNodesetID string
+		var nodeset Nodeset
+		for {
+			row, err := input.Next()
+			if err == io.EOF {
+				if currNodesetID != "" {
+					ret[currNodesetID] = nodeset
+				} else {
+					processed = false
+				}
+				currNodesetID = ""
+				break
+			}
+			nodesetId := getColumn(row, "nodeset_id")
+			if _, seen := ret[nodesetId]; seen {
+				continue
+			}
+			currNodesetID = nodesetId
+			nodeset = Nodeset{
+				ID:     currNodesetID,
+				Labels: make([]string, 0),
+				Data:   make(map[string]NodesetData),
+			}
+			// Process a new nodeset, parse row and add properties nodeset
+			parseRow := func(row []string) error {
+				var nsDataID string
+				for idx := 0; idx < len(input.ColumnNames()); idx++ {
+					val := row[idx]
+					header := input.ColumnNames()[idx]
+					if header == "nodeset_id" {
+						continue
+					} else if header == "concept_id" {
+						nsDataID = val
+						if _, seen := nodeset.Data[nsDataID]; !seen {
+							nodeset.Data[nsDataID] = NodesetData{
+								ID:         nsDataID,
+								Labels:     make([]string, 0),
+								Properties: make(map[string]interface{}),
+							}
+						}
+					} else if header == "node_labels" {
+						if nsl, ok := nodeset.Data[nsDataID]; ok {
+							nsl.Labels = append(nsl.Labels, strings.Fields(row[idx])...)
+							nodeset.Data[nsDataID] = nsl
+						}
+					} else {
+						props, _ := nodeset.Data[nsDataID].Properties[input.ColumnNames()[idx]].([]string)
+						props = append(props, val)
+						nodeset.Data[nsDataID].Properties[input.ColumnNames()[idx]] = props
+					}
+				}
+				return nil
+			}
+			if err := parseRow(row); err != nil {
+				return nil, err
+			}
+			processed = true
+			for {
+				r, err := input.Next()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return nil, err
+				}
+				// continue scanning with only with the same nodesetID
+				if currNodesetID != "" && nodesetId != currNodesetID {
+					continue
+				}
+				if err := parseRow(r); err != nil {
+					return nil, err
+				}
+			}
+		}
+		if !processed {
+			break
+		}
+	}
+	return ret, nil
+}
+
+// func buildNodesetNodes(nodeset Nodeset) []*lpg.Node {
+// 	grph := ls.NewDocumentGraph()
+// 	nodes := make([]*lpg.Node, 0)
+// 	for _, nsD := range nodeset.Data {
+// 		nodes = append(nodes, grph.NewNode(nsD.Labels, nsD.Properties))
+// 	}
+// 	return nodes
+// }
+
+// oldNodeset is nodeset pulled from DB
+func diff(oldNodeset, newNodeset Nodeset) (insertions []NodesetData, deletions []string, updates []NodesetData) {
+	// compare roots
+	if !lpg.NewStringSet(oldNodeset.Labels...).IsEqual(lpg.NewStringSet(newNodeset.Labels...)) {
```

Comment:
Diff needs to return a flag to describe what needs to be done with the root node

If oldNodeset.ID=="", return insertRoot
if labels or props changed at root, return updateRoot
if newNodeset.ID="", return deleteRoot